js/iframe-coordinator/src/messages/ClientToHost.ts

```diff
@@ -7,14 +16,42 @@ export type ClientToHost =
   | LabeledToast
   | LabeledNavRequest;
 
-export function validate(msg: ClientToHost): ClientToHost {
-  // TODO: actually validate cases
-  switch (msg.msgType) {
-    case 'navRequest':
-      return msg;
-    case 'toastRequest':
-      return msg;
-    case 'publish':
-      return msg;
+const labeledPublicationDecoder = object({
+  msgType: constant('publish'),
+  msg: object({
+    topic: string,
+    payload: mixed
+  })
+});
+
+const labeledToastDecoder = object({
+  msgType: constant('toastRequest'),
+  msg: object({
+    title: optional(string),
+    message: string,
+    custom: mixed
+  })
+});
+
+const labeledNavRequestDecoder = object({
+  msgType: constant('navRequest'),
+  msg: object({
+    url: string
+  })
+});
+
+const decoder = guard(
+  either3(
+    labeledPublicationDecoder,
+    labeledToastDecoder,
+    labeledNavRequestDecoder
+  )
+);
```

Comment:
I think we can resolve the label decoding issue by not using the `either` decode option and instead using the switch on message type to just decode/validate the `msg` portion of the incoming message.  The switch statement has effectively already checked that the label was valid and known, so all we need to do at that point is decode/validate the message data.

Reply:
I had thought about that after your comments.  It does mean that they could potentially send other properties that aren't being validated.  But I guess we could always do that part manually as well, if we really wanted to block exactness.  And if we did ever expand the root level to include another property it does make it more difficult to validate.

But I guess, those are acceptable compromises.  I can put up the changes tomorrow.

Reply:
I think it's fine if they're sending extra keys we don't use. That can be a nice tool for version compatibility eventually.